tests/ui/abi/transparent.rs

```diff
@@ -0,0 +1,52 @@
+#![feature(rustc_attrs)]
+#![crate_type = "lib"]
+
+// Check that various `transparent` wrappers result in equal ABIs.
+#[repr(transparent)]
+struct Wrapper1<T>(T);
+#[repr(transparent)]
+struct Wrapper2<T>(T, ());
+#[repr(transparent)]
+struct Wrapper3<T>(T, [u8; 0]);
+
+macro_rules! test_transparent {
+    ($name:ident, $t:ty) => {
+        mod $name {
+            #[rustc_abi(assert_eq)]
+            type Test1 = (extern "C" fn($t), extern "C" fn(super::Wrapper1<$t>));
+            //~^ ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            #[rustc_abi(assert_eq)]
+            type Test2 = (extern "C" fn($t), extern "C" fn(super::Wrapper2<$t>));
+            //~^ ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            #[rustc_abi(assert_eq)]
+            type Test3 = (extern "C" fn($t), extern "C" fn(super::Wrapper3<$t>));
+            //~^ ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+            //~| ERROR: ABIs are (sufficiently) equal
+        }
+    }
+}
+
+test_transparent!(simple, i32);
+test_transparent!(reference, &'static i32);
+test_transparent!(array, [i32; 8]);
+test_transparent!(pair, (i32, f32));
+test_transparent!(tuple, (i32, f32, i64, f64));
+test_transparent!(zst, ());
+test_transparent!(empty_array, [u32; 0]);
```

Comment:
I expect this will fail on some targets due to https://github.com/rust-lang/rust/issues/115336. I'm not sure which targets we run the test suite on, though.